pkg/cmd/step/create/step_create_meta_pipeline.go

```diff
@@ -0,0 +1,325 @@
+	cmd.Flags().StringArrayVarP(&options.CustomEnvs, "env", "e", nil, "List of custom environment variables to be applied to resources that are created (can be use multiple times)")
+	// TODO issue #3342 what is the long term strategy here? Should this be handled via env variables? Are CLI flags a good idea? What's about build reproducibility
+	cmd.Flags().StringVar(&options.DefaultImage, "default-image", syntax.DefaultContainerImage, "Specify the docker image to use if there is no image specified for a step")
+
+	cmd.Flags().StringVarP(&options.Context, "context", "c", "", "The pipeline context if there are multiple separate pipelines for a given branch")
+	cmd.Flags().StringVar(&options.Revision, "revision", "", "The git revision to checkout, can be a branch name or git sha")
+	cmd.Flags().StringVar(&options.PullRequestNumber, "pr-number", "", "If a Pull Request this is it's number")
+	cmd.Flags().StringVar(&options.SourceName, "source", "source", "The name under which to checkout the repository. Defaults to 'source'")
+	cmd.Flags().StringVarP(&options.Trigger, "trigger", "t", string(pipelineapi.PipelineTriggerTypeManual), "The kind of pipeline trigger")
+	cmd.Flags().StringVar(&options.ServiceAccount, "service-account", "tekton-bot", "The Kubernetes ServiceAccount to use to run the pipeline")
+
+	cmd.Flags().StringVarP(&options.OutDir, "output", "o", "out", "The directory to write the output to as YAML")
+	cmd.Flags().BoolVar(&options.NoApply, "no-apply", false, "Disables creating the pipeline resources in the kubernetes cluster and just outputs the generated Task to the console or output file")
+	cmd.Flags().BoolVar(&options.ViewSteps, "view", false, "Just view the steps that would be created")
+
+	options.AddCommonFlags(cmd)
+	return cmd
+}
+
+// Run implements this command
+func (o *StepCreatePipelineOptions) Run() error {
+	err := o.validateCommandLineFlags()
+	if err != nil {
+		return err
+	}
+
+	gitInfo, _ := gits.ParseGitURL(o.SourceURL)
+
+	tektonClient, jxClient, kubeClient, ns, err := o.getClientsAndNamespace()
+	if err != nil {
+		return err
+	}
+
+	podTemplates, err := o.getPodTemplates(kubeClient, ns, apps.AppPodTemplateName)
+	if err != nil {
+		return errors.Wrap(err, "unable to retrieve pod templates")
+	}
+
+	pipelineResourceName := tekton.PipelineResourceName(gitInfo, o.Branch, o.Context)
+	buildNumber, err := tekton.GenerateNextBuildNumber(tektonClient, jxClient, ns, gitInfo, o.Branch, retryDuration, pipelineResourceName)
+	if err != nil {
+		return errors.Wrap(err, "unable to determine next build number")
+	}
+
+	if o.Verbose {
+		log.Logger().Infof("creating meta pipeline CRDs for build %s of repository '%s/%s'", buildNumber, gitInfo.Organisation, gitInfo.Name)
+	}
+	extendingApps, err := metapipeline.GetExtendingApps(jxClient, ns)
+	if err != nil {
+		return err
+	}
+
+	pipelineName := tekton.PipelineResourceName(gitInfo, o.Branch, o.Context)
```

Comment:
So this is an interesting thing - we probably do actually want the same `PipelineResource` name for both the metapipeline and the generated pipeline, but we need a different *name* for the `Pipeline` (and `PipelineRun`, `Task`, `PipelineStructure`). Otherwise we end up with dupes. So we should probably pass in both a `PipelineName` and a `MetaPipelineName`, both of which need to follow the `kube.ToValidTruncatedName` rules.

Reply:
Right, this is something I'd like to discuss. I can show you how it works out atm.